src/test/java/software/amazon/encryption/s3/S3AsyncEncryptionClientTest.java

```diff
@@ -67,6 +79,240 @@ public static void setUp() throws NoSuchAlgorithmException {
+                .region(Region.of(KMS_REGION.toString()))
+                .kmsKeyId(KMS_KEY_ID)
+                .build();
+
+        final String input = "SimpleTestOfV3EncryptionClientAsync";
+
+        s3Client.putObject(builder -> builder
+                        .bucket(BUCKET)
+                        .key(objectKey)
+                        .build(),
+                AsyncRequestBody.fromString(input)).join();
+
+        ResponseBytes<GetObjectResponse> objectResponse = s3Client.getObject(builder -> builder
+                .bucket(BUCKET)
+                .key(objectKey)
+                .build(), AsyncResponseTransformer.toBytes()).join();
+        String output = objectResponse.asUtf8String();
+        assertEquals(input, output);
+
+        // Cleanup
+        deleteObject(BUCKET, objectKey, s3Client);
+        s3Client.close();
+    }
+
+    @Test
+    public void s3AsyncEncryptionClientTopLevelAlternateCredentials() {
+        final String objectKey = appendTestSuffix("wrapped-s3-async-client-with-top-level-alternate-credentials");
+        final String input = "S3EncryptionClientTopLevelAlternateCredsTest";
+
+        // use alternate creds
+        AwsCredentialsProvider creds = new S3EncryptionClientTestResources.AlternateRoleCredentialsProvider();
+
+        S3AsyncClient s3Client = S3AsyncEncryptionClient.builder()
+                .credentialsProvider(creds)
+                .region(Region.of(KMS_REGION.toString()))
+                .kmsKeyId(KMS_KEY_ID)
+                .build();
+
+        // using the original key fails
+        try {
+            s3Client.putObject(builder -> builder
+                            .bucket(BUCKET)
+                            .key(objectKey)
+                            .build(),
+                    AsyncRequestBody.fromString(input)).join();
+            fail("expected exception");
+        } catch (KmsException exception) {
+            // expected
+            assertTrue(exception.getMessage().contains("is not authorized to perform"));
+        } finally {
+            s3Client.close();
+        }
+
+        // using the alternate key succeeds
+        S3AsyncClient s3ClientAltCreds = S3AsyncEncryptionClient.builder()
+                .credentialsProvider(creds)
+                .region(Region.of(KMS_REGION.toString()))
+                .kmsKeyId(ALTERNATE_KMS_KEY)
+                .build();
+
+        s3ClientAltCreds.putObject(builder -> builder
+                        .bucket(BUCKET)
+                        .key(objectKey)
+                        .build(),
+                AsyncRequestBody.fromString(input)).join();
+
+        ResponseBytes<GetObjectResponse> objectResponse = s3ClientAltCreds.getObject(builder -> builder
+                .bucket(BUCKET)
+                .key(objectKey)
+                .build(), AsyncResponseTransformer.toBytes()).join();
+        String output = objectResponse.asUtf8String();
+        assertEquals(input, output);
+
+        // Cleanup
```

Comment:
(nit)
I like putting the cleanup at the beginning of tests
If a test fails, the cleanup won't happen since it happens at the end of the test

Reply:
The `appendTestSuffix` method ensures that object keys aren't reused. In the case of test failure, objects are cleaned up via the lifecycle rule configuration on the S3 bucket.